datadog/resource_datadog_metric_tag_configuration.go

```diff
@@ -0,0 +1,299 @@
+package datadog
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	datadogV2 "github.com/DataDog/datadog-api-client-go/api/v2/datadog"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/validators"
+)
+
+func resourceDatadogMetricTagConfiguration() *schema.Resource {
+	return &schema.Resource{
+		Description: "Provides a Datadog metric tag configuration resource. This can be used to modify tag configurations for metrics.",
+		Create:      resourceDatadogMetricTagConfigurationCreate,
+		Read:        resourceDatadogMetricTagConfigurationRead,
+		Update:      resourceDatadogMetricTagConfigurationUpdate,
+		Delete:      resourceDatadogMetricTagConfigurationDelete,
+		CustomizeDiff: func(diff *schema.ResourceDiff, meta interface{}) error {
+			_, includePercentilesOk := diff.GetOkExists("include_percentiles")
+			if !includePercentilesOk {
+				// if there was no change to include_percentiles we don't need special handling
+				return nil
+			}
+
+			metricType, ok := diff.GetOkExists("metric_type")
+			if !ok {
+				// no change to metric_type so no special handling
+				return nil
+			}
+			metricTypeValidated, err := datadogV2.NewMetricTagConfigurationMetricTypesFromValue(metricType.(string))
+			if err != nil {
+				return utils.TranslateClientError(err, "error validating diff")
+			}
+			if includePercentilesOk && *metricTypeValidated != datadogV2.METRICTAGCONFIGURATIONMETRICTYPES_DISTRIBUTION {
+				return fmt.Errorf("cannot use include_percentiles with a metric_type of %s, must use metric_type of 'distribution'", metricType)
+			}
+
+			return nil
+		},
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"metric_name": {
+				Description: "The metric name for this resource.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+				ValidateFunc: func(val interface{}, k string) (warns []string, errs []error) {
+					v := val.(string)
+					regExpWarns, regExpErrs := validation.StringMatch(regexp.MustCompile(`^[A-Za-z][A-Za-z0-9\\.\\_]*$`), "metric name must be valid")(v, k)
+					lenWarns, lenErrs := validation.StringLenBetween(1, 200)(v, k)
```

Comment:
I think you can use https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/validation#All and remove most of the code in there.